Objects/call.c

```diff
@@ -167,6 +167,38 @@ PyObject_VectorcallDict(PyObject *callable, PyObject *const *args,
     return _PyObject_FastCallDictTstate(tstate, callable, args, nargsf, kwargs);
 }
 
+static void
+object_is_not_callable(PyThreadState *tstate, PyObject *callable)
+{
+    if (Py_IS_TYPE(callable, &PyModule_Type)) {
+        // >>> import pprint
+        // >>> pprint(thing)
+        // Traceback (most recent call last):
+        //   File "<stdin>", line 1, in <module>
+        // TypeError: 'module' object is not callable. Did you mean: 'pprint.pprint(...)'?
+        PyObject *name = PyModule_GetNameObject(callable);
+        if (name == NULL) {
```

Comment:
Should we instead suppress the new exception here and go back to the "object is not callable" message? Looks like this can happen if someone deletes the `__name__` entry from a module. It would be confusing if you then get an error "module has no attribute __name__" when you try to call the module.

Reply:
Yeah, that makes sense.

Reply:
It seems like this could be

```c
PyObject *name = PyModule_GetNameObject(callable);
PyObject *attr = NULL;
_PyErr_Format(tstate, PyExc_TypeError, "'%.200s' object is not callable",
              Py_TYPE(callable)->tp_name);

if (name != NULL) {
    int res = _PyObject_LookupAttr(callable, name, &attr);
    if (res > 0 && PyCallable_Check(attr)) {
        _PyErr_Format(tstate, PyExc_TypeError,
                      "'%.200s' object is not callable. "
                      "Did you mean: '%U.%U(...)'?",
                      Py_TYPE(callable)->tp_name, name, name);
    }
}
Py_XDECREF(attr);
Py_XDECREF(name);
```

Reply:
Not an expert on exceptions, but does `_PyErr_Format()` just replace the previous Exception set?

Reply:
Oh wait, that's wrong, a minute.

Reply:
This could get rid of the multiple appearance of Py_DECREF, the early return and the label

Reply:
Brandt and I discussed this in person. Your suggestion should indeed work; PyErr_Format will clear the previous exception before setting a new one. However, I think this makes the code harder to understand than the current behavior of conditionally setting just one exception.

Reply:
This was to avoid early return which often considered as a non-desirable behavior, and to avoid redundant code. The logic of the current implementation is a bit nested. The key idea of the improvement is - if certain condition is met, use the better error, otherwise fallback to the original.

Actually it could be:

```c
PyObject *name = PyModule_GetNameObject(callable);
PyObject *attr = NULL;
if (name != NULL
        && _PyObject_LookupAttr(callable, name, &attr) > 0
        && PyCallable_Check(attr)) { 
    _PyErr_Format(tstate, PyExc_TypeError,
                  "'%.200s' object is not callable. "
                  "Did you mean: '%U.%U(...)'?",
                  Py_TYPE(callable)->tp_name, name, name);
} else {
    _PyErr_Format(tstate, PyExc_TypeError,
                  "'%.200s' object is not callable",
                  Py_TYPE(callable)->tp_name);
}
Py_XDECREF(attr);
Py_XDECREF(name);
```

Reply:
This new example still "overwrites" the error if `_PyObject_LookupAttr` raises an exception. Again, your code is a bit simpler in structure, but I still find it much easier to reason about the correctness of the current form.

This variant also *always* calls `PyModule_GetNameObject` on `callable`, which means this will raise and overwrite an error for non-modules, too.

Reply:
Sorry I just omitted the Module check as I thought that will always be there.

Reply:
But then we would still either need to duplicate the "normal" `TypeError` raise, or use a label (or flag) like this PR does. Right?

Reply:
The current implementation effectively "overwrites" the error if any exception is raised. We could add `_PyErr_Clear(tstate);` before setting the fallback error message to explicitly indicate "we don't care what exception is there, I'll just clear it". It has no effect when no exception is set.

I think of this issue as a simple check - if certain condition is met, raise this error, otherwise that error. No exception during this process matters.

When I take a look at the current implementation, I had to go though every logic path to confirm that the reference count was correct. That's another aspect that needs to be "reason about".

Reply:
> But then we would still either need to duplicate the "normal" `TypeError` raise, or use a label (or flag) like this PR does. Right?

Ah, that's something I missed. I would suggest this but it might be considered more difficult to reason about:

```c
PyObject *name = NULL;
PyObject *attr = NULL;
if (Py_IS_TYPE(callable, &PyModule_Type)
        && (name = PyModule_GetNameObject(callable)) 
        && _PyObject_LookupAttr(callable, name, &attr) > 0
        && PyCallable_Check(attr)) {
    ...
}
```

Reply:
Do you like the walrus operator? :)